more fixes

neheb opened the pull request.

piponazo commented on src/crwimage_int.cpp:
```diff
@@ -37,7 +37,7 @@ constexpr RotationMap::OmList RotationMap::omList_[] = {
 uint16_t RotationMap::orientation(int32_t degrees) {
   uint16_t o = 1;
-  for (auto&& [deg, orient] : omList_) {
+  for (auto&& [orient, deg] : omList_) {
```
Was this a bug? I am surprised that this is not changing behaviour 😮

neheb replied:
Yeah it was. Maybe this code is not tested

piponazo commented on app/actions.cpp:
```diff
@@ -1290,9 +1290,9 @@ void Modify::delMetadatum(Exiv2::Image* pImage, const ModifyCmd& modifyCmd) {
     const Exiv2::XmpKey xmpKey(modifyCmd.key_);
-    if ((pos = xmpData.findKey(xmpKey)) != xmpData.end()) {
+    auto pos = xmpData.findKey(xmpKey);
+    if (pos != xmpData.end()) {
```
[suggestion] You can still use the C++17 feature to declare the variable inside the if and afterwards do the comparison.

neheb replied:
I know about it. Doesn't look as nice IMO

neheb committed:
replace localtime with _s/_r variant

cppcheck warns on localtime which is not necessarily threadsafe.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
fix memory leak

Found with cppcheck.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use upper() function

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
make conversions explicit

Found with MSVC's C4244

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
explicit conversion

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
fix wrong snprintf format

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
replace array with vector

Using the constructor is one less line. Plus it seems to fix bugs in
Apple's clang.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use count_if

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use std::replace

shorter

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
Remaining warnings. Not too sure about them.

```
2023-02-12T02:40:03.8824101Z D:\a\exiv2\exiv2\src\basicio.cpp(684) : warning C6308: 'realloc' might return null pointer: assigning null pointer to 'data_', which is passed as an argument to 'realloc', will cause the original memory block to be leaked.
2023-02-12T02:40:03.8824996Z D:\a\exiv2\exiv2\src\basicio.cpp(941) : warning C6262: Function uses '103332' bytes of stack.  Consider moving some data to heap.
2023-02-12T02:40:10.1230140Z D:\a\exiv2\exiv2\src\basicio.cpp(412) : warning C4701: potentially uninitialized local variable 'buf2' used
2023-02-12T02:41:00.8113062Z ../src/http.cpp(212): warning C4244: 'argument': conversion from 'int' to 'u_short', possible loss of data
2023-02-12T02:41:00.9206261Z D:\a\exiv2\exiv2\src\http.cpp(133) : warning C6262: Function uses '33996' bytes of stack.  Consider moving some data to heap.
2023-02-12T02:41:01.0155752Z D:\a\exiv2\exiv2\src\http.cpp(151) : warning C6031: Return value ignored: 'WSAStartup'.
2023-02-12T02:41:08.0309475Z D:\a\exiv2\exiv2\src\http.cpp(240) : warning C6386: Buffer overrun while writing to 'buffer'.: Lines: 239, 134, 136, 138, 140, 143, 144, 145, 150, 151, 154, 155, 156, 157, 158, 159, 161, 162, 163, 166, 167, 168, 169, 170, 171, 172, 175, 176, 177, 178, 179, 180, 181, 184, 187, 192, 194, 199, 200, 204, 207, 208, 210, 211, 212, 216, 225, 229, 230, 234, 235, 239, 240
2023-02-12T02:41:39.5074921Z C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\include\xutility(4117): warning C4244: '=': conversion from 'uint32_t' to 'Exiv2::byte', possible loss of data
2023-02-12T02:42:00.5230007Z D:\a\exiv2\exiv2\src\properties.cpp(5178) : warning C6031: Return value ignored: 'std::unique_ptr<Exiv2::XmpKey::Impl,std::default_delete<Exiv2::XmpKey::Impl> >::->'.
2023-02-12T02:42:00.6012957Z D:\a\exiv2\exiv2\src\properties.cpp(5182) : warning C6031: Return value ignored: 'std::unique_ptr<Exiv2::XmpKey::Impl,std::default_delete<Exiv2::XmpKey::Impl> >::->'.
2023-02-12T02:42:18.7971400Z D:\a\exiv2\exiv2\src\quicktimevideo.cpp(849) : warning C6011: Dereferencing NULL pointer 'tv'. : Lines: 798, 799, 800, 802, 803, 804, 805, 807, 808, 809, 810, 811, 813, 814, 816, 818, 820, 822, 825, 828, 831, 835, 840, 841, 842, 843, 844, 846, 849
2023-02-12T02:42:19.0981220Z C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.34.31933\include\utility(172): warning C4244: 'initializing': conversion from '_Ty' to '_Ty1', possible loss of data
2023-02-12T02:42:50.8094181Z D:\a\exiv2\exiv2\src\version.cpp(220) : warning C6240: (<expression> && <non-zero constant>) always evaluates to the result of <expression>:  Did you intend to use the bitwise-and (&) operator? If not, consider removing the redundant '<non-zero constant>' and the && operator.
2023-02-12T02:42:50.8100266Z D:\a\exiv2\exiv2\src\version.cpp(222) : warning C6240: (<expression> && <non-zero constant>) always evaluates to the result of <expression>:  Did you intend to use the bitwise-and (&) operator? If not, consider removing the redundant '<non-zero constant>' and the && operator.
2023-02-12T02:42:51.0751153Z D:\a\exiv2\exiv2\src\version.cpp(220) : warning C6326: Potential comparison of a constant with another constant.
2023-02-12T02:43:00.4190440Z D:\a\exiv2\exiv2\src\version.cpp(222) : warning C6326: Potential comparison of a constant with another constant.
2023-02-12T02:43:15.7764035Z D:\a\exiv2\exiv2\src\xmpsidecar.cpp(39) : warning C6262: Function uses '65700' bytes of stack.  Consider moving some data to heap.
2023-02-12T02:44:28.3502423Z D:\a\exiv2\exiv2\src\pentaxmn_int.cpp(909) : warning C6297: Arithmetic overflow.  Results might not be an expected value.
```

kevinbackhouse commented on src/image.cpp:
```diff
@@ -469,9 +469,9 @@ void Image::printIFDStructure(BasicIo& io, std::ostream& out, Exiv2::PrintStruct
-            DataBuf bytes(byteslen);                                                  // allocate a buffer
-            io.readOrThrow(bytes.data(), byteslen, ErrorCode::kerCorruptedMetadata);  // read
-            MemIo memIo(bytes.c_data(), byteslen);                                    // create a file
+            auto b = DataBuf(byteslen);                                           // allocate a buffer
```
What's the benefit of doing it this way?

kevinbackhouse commented on src/preview.cpp:
```diff
@@ -861,21 +860,15 @@ DataBuf decodeHex(const byte* src, size_t srcSize) {
-    if (decodeBase64Table[static_cast<unsigned char>(src[srcPos])] != invalid)
-      validSrcSize++;
-  }
+  auto validSrcSize = static_cast<unsigned long>(
```
Can we use `size_t` here? I'd like to eliminate `long` from the codebase.
